dexios/src/domain/pack.rs

```diff
@@ -0,0 +1,245 @@
+use std::cell::RefCell;
+use std::io::{BufReader, BufWriter, Cursor, Read, Seek, Write};
+
+use dexios_core::header::{HashingAlgorithm, HeaderType};
+use dexios_core::primitives::BLOCK_SIZE;
+use dexios_core::protected::Protected;
+use zip::write::FileOptions;
+
+use crate::domain;
+
+#[derive(Debug)]
+pub enum Error {
+    AddDirToArchive,
+    AddFileToArchive,
+    FinishArchive,
+    ReadData,
+    WriteData,
+    Encrypt(domain::encrypt::Error),
+    Overwrite(domain::overwrite::Error),
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use Error::*;
+        match self {
+            AddDirToArchive => f.write_str("Unable to add directory to archive"),
+            AddFileToArchive => f.write_str("Unable to add file to archive"),
+            FinishArchive => f.write_str("Unable to finish archive"),
+            ReadData => f.write_str("Unable to read data"),
+            WriteData => f.write_str("Unable to write data"),
+            Encrypt(inner) => write!(f, "Unable to encrypt archive: {}", inner),
+            Overwrite(inner) => write!(f, "Unable to overwrite archive: {}", inner),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+pub struct Request<'a, RW>
+where
+    RW: Read + Write + Seek,
+{
+    pub writer: &'a RefCell<RW>,
+    pub compress_files: Vec<domain::storage::File<RW>>,
+    pub compression_method: zip::CompressionMethod,
+    pub header_writer: Option<&'a RefCell<RW>>,
+    pub raw_key: Protected<Vec<u8>>,
+    // TODO: don't use external types in logic
+    pub header_type: HeaderType,
+    pub hashing_algorithm: HashingAlgorithm,
+}
+
+pub fn execute<RW>(req: Request<RW>) -> Result<(), Error>
+where
+    RW: Read + Write + Seek,
+{
+    // 1. Create in memory zip archive.
+    let mut zip_content = vec![];
+    let file = Cursor::new(&mut zip_content);
+    let mut zip_writer = zip::ZipWriter::new(BufWriter::new(file));
```

Comment:
@brxken128  What do you think about this? Or do you still create a temporary file for the archive?

Reply:
Clear) I'll return with changes later)

Reply:
It would be ideal to have the following:

```rust
let cursor = io::Cursor::new(data);
let file = io::File::open(file_path);
let zip = zip::ZipWriter::new(buffer);
let encrypt = dexios::EncryptWriter::new(
    cursor/* or file or zip*/,
    options /* header_type, algorithm and etc*/
);

// EncryptWriter has `io::Write`;
encrypt.write_all("hello world")?; // write data to file / cursor / zip and then encrypt this part.

encrypt.finish()?; // add the header to the output file and flush result.
encrypt.finish_detached(header_writer)? // add the header to another file / cursor and etc.

```

Reply:
I can try to play with it)

Reply:
> It would be ideal to have the following:
> 
> ```rust
> let cursor = io::Cursor::new(data);
> let file = io::File::open(file_path);
> let zip = zip::ZipWriter::new(buffer);
> let encrypt = dexios::EncryptWriter::new(
>     cursor/* or file or zip*/,
>     options /* header_type, algorithm and etc*/
> );
> 
> // EncryptWriter has `io::Write`;
> encrypt.write_all("hello world")?; // encrypt input data and then write data to file / cursor / zip
> 
> encrypt.finish()?; // add the header to the output file and flush result.
> encrypt.finish_detached(header_writer)? // add the header to another file / cursor and etc.
> ```

That's probably the best solution to this actually, just write encrypted data in place of the files and use the same header for each. Easier than fully manipulating zip formats and such, but still beyond adequate in terms of functionality and performance. Nice thinking!

Reply:
I think you don't understand my idea)

Reply:
> My idea was that we would write data incrementally through the pipes (EncryptWriter -> File). From the top level (EncryptWriter) to the lower level (File / Cursor). Maybe with a zip archive it will be more complicated, but we can try to start with a Cursor.
> 
> I think it should be very performant. Because we will encrypt the chunk, then write and then delete it from memory.

I'm sure we can make it work - it sounds perfect on paper :)